x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,352 @@
+                        "Skipping security auto configuration because it appears that TLS is already configured.");
+                return;
+            }
+            // Silently skipping security auto configuration because file realm is configured
+            // Auto-configuration only requires that the file realm be enabled, and, for optimal experience,
+            // also be the first in the chain.
+            if (false == env.settings().getByPrefix(RealmSettings.realmSettingPrefix(FileRealmSettings.TYPE)).isEmpty()) {
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because it appears that a file-based realm is already configured.");
+                return;
+            }
+        }
+        final Path ymlPath = env.configFile().resolve("elasticsearch.yml");
+        // The node's yml configuration is mounted read-only, which is a sign that the configuration mustn't change.
+        // Inform that auto-configuration will not run.
+        if (false == Files.isWritable(ymlPath)) {
+            terminal.println(Terminal.Verbosity.NORMAL, String.format(Locale.ROOT, "Skipping security auto configuration because " +
+                    "the configuration file [%s] is not writable", ymlPath));
+            return;
+        }
+        // Best effort determine if the node's keystore is writable.
+        // Inform that auto-configuration will not run if keystore cannot be updated.
+        final Path keystorePath = env.configFile().resolve(KeyStoreWrapper.KEYSTORE_FILENAME);
+        if (Files.exists(keystorePath) && false == Files.isWritable(keystorePath)) {
+            terminal.println(Terminal.Verbosity.NORMAL, String.format(Locale.ROOT, "Skipping security auto configuration because " +
+                    "the node keystore file [%s] is not writable", keystorePath));
+            return;
+        }
+
+        // TODO round/truncate to day, as it is easier on the tests
+        //Rounding rounding = new Rounding.Builder(Rounding.DateTimeUnit.DAY_OF_MONTH).timeZone(ZoneOffset.UTC).build();
+        //ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC).truncatedTo(ChronoUnit.DAYS);
+        final ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC);
+        final String instantAutoConfigName = "auto_generated_on_" + autoConfigDate.toInstant().getEpochSecond();
+        final Path instantAutoConfigDir = env.configFile().resolve(instantAutoConfigName);
+        try {
+            Files.createDirectory(instantAutoConfigDir);
+        } catch (IOException e) {
+            throw new UserException(ExitCodes.CANT_CREATE, "Could not create auto configuration directory", e);
+        }
+
+        // the transport key-pair is the same across the cluster and is trusted without hostname verification (it is self-signed),
+        // do not populate the certificate's IP, DN, and CN certificate fields
+        // TODO is CN appropriate here?
```

Comment:
Yes, a principal is always appropriate

Reply:
I meant if CN=hostname is OK, ie how to "transform" a hostname or FQDN to a Distinguished Name. I ended up using :
```
    /**
     * See RFC 2247 Using Domains in LDAP/X.500 Distinguished Names
     * @param domain active directory domain name
     * @return LDAP DN, distinguished name, of the root of the domain
     */
    public static String buildDnFromDomain(String domain) {
        return "DC=" + domain.replace(".", ",DC=");
    }
```

Reply:
I think it doesn't _really_ matter . More and more implementations do not even check the CN in any way and depend on SANs. But +1 to be as correct as possible